### PR TITLE
Remove the "with-exceptions" backend from the compiler

### PR DESCRIPTION
**Pull request by AltGr**
Also, some debugging of the monomorphisation pass.

Quoting the main patch from this PR:

```quote
  *Disclaimer*: This is intended for discussion

  My impression is that the with-exceptions backend is to be superseded by the
  without-exception backend, which is more general and more efficient. Therefore,
  seeing the added complexity of maintaining the two in parallel, I see no good
  reason to keep the with-exceptions version now that the equivalence of their
  semantics have been proved.

  It will also be nice to reduce divergences between the different backends ; and
  this should make further simplifications possible (e.g. some thunkings may no
  longer be needed)

  Of course I am ready to hear arguments in favor of keeping it, be it in the mid-
  or long-term.
```

The PR also changes the compilation of default terms to rely on a simpler, lower-level `HandleExceptions` operator that just reduces a list of options, wrapped in a `EMatch` block; operators without function arguments are easier to manipulate, and this makes the evaluation strategy clear without having to rely on thunking (i.e. functions of unit).

### CODE REVIEW

**Comment by AltGr** [comment]
Added sister branch in catala-examples (to remove the flag that this made obsolete).
Don't forget to merge both in sync!

File: compiler/scalc/from_lcalc.ml [hunk 1]
```diff
@@ -138,15 +131,15 @@ and translate_expr (ctxt : 'm ctxt) (expr : 'm L.expr) : RevBlock.t * A.expr =
...
-          op = Op.HandleDefaultOpt, _;
-          args = [_exceptions; _just; _cons];
-          tys = _;
-        }
-      when ctxt.config.keep_special_ops ->
-      (* This should be translated as a statement *)
-      raise (NotAnExpr { needs_a_local_decl = true })
+    (* | EAppOp
+     *     {
+     *       op = Op.HandleDefaultOpt, _;
```
Comment (denismerigoux) [inline compiler/scalc/from_lcalc.ml:136]: Leftover comment?

File: compiler/scalc/to_c.ml [hunk 2]
```diff
@@ -368,8 +367,8 @@ let rec format_expression (ctx : decl_ctx) (fmt : Format.formatter) (e : expr) :
     Format.fprintf fmt "%a %a" format_op op (format_expression ctx) arg1
   | EAppOp { op; args = [arg1] } ->
     Format.fprintf fmt "%a(%a)" format_op op (format_expression ctx) arg1
-  | EAppOp { op = (HandleDefaultOpt | HandleDefault), _; args = _ } ->
-    failwith "should not happen because of keep_special_ops"
+  (* | EAppOp { op = HandleDefaultOpt, _; args = _ } ->
```
Comment (denismerigoux) [inline compiler/scalc/to_c.ml:370]: Leftover comment ?

File: compiler/scalc/to_python.ml [hunk 3]
```diff
@@ -349,27 +348,25 @@ let rec format_expression ctx (fmt : Format.formatter) (e : expr) : unit =
...
-      "%a(@[<hov 0>SourcePosition(filename=\"%s\",@ start_line=%d,@ \
-       start_column=%d,@ end_line=%d, end_column=%d,@ law_headings=%a), %a)@]"
-      format_op op (Pos.get_file pos) (Pos.get_start_line pos)
-      (Pos.get_start_column pos) (Pos.get_end_line pos) (Pos.get_end_column pos)
-      format_string_list (Pos.get_law_info pos)
+  (* | EAppOp { op = ((HandleDefaultOpt), _) as op; args } ->
+   *   let pos = Mark.get e in
+   *   Format.fprintf fmt
+   *     "%a(@[<hov 0>SourcePosition(filename=\"%s\",@ start_line=%d,@ \
+   *      start_column=%d,@ end_line=%d, end_column=%d,@ law_headings=%a), %a)@]"
```
Comment (denismerigoux) [inline compiler/scalc/to_python.ml:355]: Leftover comment ?

File: compiler/scalc/to_r.ml [hunk 4]
```diff
@@ -321,24 +320,24 @@ let rec format_expression (ctx : decl_ctx) (fmt : Format.formatter) (e : expr) :
...
-      format_string_list (Pos.get_law_info pos)
-      (Format.pp_print_list
-         ~pp_sep:(fun fmt () -> Format.fprintf fmt ",@ ")
-         (format_expression ctx))
-      args
+  (* TODO: port the following to avoid-exceptions
+   * | EAppOp { op = (HandleDefault as op), _; args; _ } ->
+   *   let pos = Mark.get e in
+   *   Format.fprintf fmt
+   *     "%a(@[<hov 0>catala_position(filename=\"%s\",@ start_line=%d,@ \
```
Comment (denismerigoux) [inline compiler/scalc/to_r.ml:330]: Leftover comment ?

File: runtimes/r/R/runtime.R [hunk 5]
```diff
@@ -363,36 +363,6 @@ catala_assertion_failure <- function(pos) {
 
 ################ Defaults #################
 
-#' @export
-catala_handle_default <- function(pos, exceptions, just, cons) {
```
Comment (denismerigoux) [inline runtimes/r/R/runtime.R:367]: Shouldn't that be replaced with a similar function to handle exceptions with options?

**Review by denismerigoux — APPROVED** [review]
Thanks @altgr!

File: compiler/lcalc/monomorphize.ml [hunk 6]
```diff
@@ -78,7 +78,7 @@ let collect_monomorphized_instances (prg : typed program) :
                           args;
                       name =
                         StructName.fresh []
-                          ( "tuple_" ^ string_of_int !option_instances_counter,
+                          ( "tuple_" ^ string_of_int !tuple_instances_counter,
```
[the remark this answers is not in the document]
Comment (AltGr) [reply compiler/lcalc/monomorphize.ml:81]: Yeah I am glad I spotted this one before we got bitten by it ;)

File: runtimes/r/R/runtime.R (hunk 5 above)
Comment (AltGr) [reply runtimes/r/R/runtime.R:367]: Indeed, that's a "TODO" in the `to_r` module that remains for later

**Comment by AltGr** [comment]
Fixed one last closure conversion bug, this should be good to go except for the actual implementation of `handle_exceptions` in R

**Pull request merged**